libs/algo/odc/algo/_tiff.py

```diff
@@ -0,0 +1,518 @@
+        return roundup16(dim)
+    return roundup16(block)
+
+
+@dataclass
+class GeoRasterInfo:
+    width: int
+    height: int
+    count: int
+    dtype: str
+    crs: str
+    transform: Affine
+    nodata: Optional[NodataType] = None
+    axis: int = 0
+
+    def gdal_opts(self):
+        out = dict(**self.__dict__)
+        if self.nodata is None:
+            out.pop("nodata")
+        out.pop("axis")
+        return out
+
+    def raster_size(self) -> int:
+        """
+        Compute raster size in bytes
+        """
+        return np.dtype(self.dtype).itemsize * self.width * self.height * self.count
+
+    @staticmethod
+    def from_xarray(xx: xr.DataArray) -> "GeoRasterInfo":
+        axis = 0
+        geobox = getattr(xx, "geobox", None)
+        if geobox is None:
+            raise ValueError("Missing .geobox on input array")
+
+        height, width = geobox.shape
+        if xx.ndim == 2:
+            count = 1
+        elif xx.ndim == 3:
+            if xx.shape[:2] == (height, width):
+                count = xx.shape[2]
+            elif xx.shape[1:] == (height, width):
+                count = xx.shape[0]
+                axis = 1
+            else:
+                raise ValueError("Geobox shape does not match array size")
+
+        nodata = getattr(xx, "nodata", None)
+
+        return GeoRasterInfo(
+            width,
+            height,
+            count,
+            xx.dtype.name,
+            str(geobox.crs),
+            geobox.transform,
+            nodata,
+            axis=axis,
+        )
+
+    def shrink2(self) -> "GeoRasterInfo":
+        return GeoRasterInfo(
+            width=half_up(self.width),
+            height=half_up(self.height),
+            count=self.count,
+            dtype=self.dtype,
+            crs=self.crs,
+            transform=self.transform * Affine.scale(2, 2),
+            nodata=self.nodata,
+        )
+
+
+class TIFFSink:
+    def __init__(
+        self,
+        info: GeoRasterInfo,
+        dst: Union[str, MemoryFile],
+        blocksize: Optional[int] = None,
+        bigtiff: Union[str, bool] = "auto",
+        lock: bool = True,
+        **extra_rio_opts,
+    ):
+        if blocksize is None:
+            blocksize = 512
+
+        if bigtiff == "auto":
+            # do bigtiff if raw raster is larger than 4GB
+            bigtiff = info.raster_size() > (1 << 32)
+
+        opts = dict(
+            driver="GTiff",
+            bigtiff=bigtiff,
+            tiled=True,
+            blockxsize=_adjust_blocksize(blocksize, info.width),
+            blockysize=_adjust_blocksize(blocksize, info.height),
+            compress="DEFLATE",
+            zlevel=6,
+            predictor=2,
+            num_threads="ALL_CPUS",
+        )
+        opts.update(info.gdal_opts())
+        opts.update(extra_rio_opts)
+
+        mem: Optional[MemoryFile] = None
+        self._mem_mine: Optional[MemoryFile] = None
+
+        if isinstance(dst, str):
+            if dst == ":mem:":
```

Comment:
I think `dst` is either set to `str(Path(temp_folder) / f"{t_name}{ext}")` or `MemoryFile(dirname=t_dir, filename=t_name + ext)` so I think this ` dst == ":mem:"` is never taken

Reply:
When coming from `save_cog` -- yes, but we might expose this class in other ways, one doesn't always need COG, normal TIFF is sometimes enough.